pinot-broker/src/main/java/org/apache/pinot/broker/routing/segmentpruner/PartitionSegmentPruner.java

```diff
@@ -54,18 +57,18 @@
  */
 public class PartitionSegmentPruner implements SegmentPruner {
   private static final Logger LOGGER = LoggerFactory.getLogger(PartitionSegmentPruner.class);
-  private static final PartitionInfo INVALID_PARTITION_INFO = new PartitionInfo(null, null);
+  private static final Map<String, PartitionInfo> INVALID_COLUMN_PARTITION_INFO_MAP = Collections.emptyMap();
 
   private final String _tableNameWithType;
-  private final String _partitionColumn;
+  private final Set<String> _partitionColumns;
   private final ZkHelixPropertyStore<ZNRecord> _propertyStore;
   private final String _segmentZKMetadataPathPrefix;
-  private final Map<String, PartitionInfo> _partitionInfoMap = new ConcurrentHashMap<>();
+  private final Map<String, Map<String, PartitionInfo>> _segmentColumnPartitionInfoMap = new ConcurrentHashMap<>();
```

Comment:
Storing a map for each segment can potentially cause performance regression and extra memory footprint, and we already observe partition pruning becomes a bottleneck when a table contains a lot of segments. Let's add it as a separate implementation so that for single partition column case (probably 99% of the use case), we don't have to pay the extra cost. You may rename the current one to `SinglePartitionColumnSegmentPruner`, then add a `MultiPartitionColumnsSegmentPruner`

Reply:
done.